### PR TITLE
Nb/feat/vision checks without indexes

### PR DESCRIPTION
**Pull request by Nadav-Barak**
Image Property Outliers, Label Property Outliers, New Labels to work without the index to image mapping

### CODE REVIEW

File: deepchecks/vision/checks/train_test_validation/new_labels.py [hunk 1]
```diff
@@ -85,10 +54,32 @@ def __init__(
...
+        labels = data.batch_to_labels(batch)
+
+        for image, label in zip(images, labels):
+            if data.task_type == TaskType.CLASSIFICATION:
+                self._update_images_dict(label, label.item(), image)
+            elif data.task_type == TaskType.OBJECT_DETECTION and len(label) > 0:
+                for class_id in np.unique(label[:, 0].numpy()):
+                    bbox_of_label = label[(label[:, 0] == class_id).nonzero().squeeze(1)]
+                    self._update_images_dict(bbox_of_label, class_id, image)
+            elif len(label) > 0:
```
Comment (noamzbr) [inline deepchecks/vision/checks/train_test_validation/new_labels.py:74]: Why not also do for segmentation? Can be separate issue.

File: deepchecks/vision/checks/train_test_validation/new_labels.py [hunk 2]
```diff
@@ -118,10 +109,10 @@ def compute(self, context: Context) -> CheckResult:
             for class_id, num_occurrences in classes_only_in_test_count.items():
                 # Create id of alphabetic characters
                 sid = ''.join([choice(string.ascii_uppercase) for _ in range(3)])
-                images_of_class_id = \
-                    list(set(test_data.classes_indices[class_id]))[:self.max_images_to_display_per_label]
-                images_combine = ''.join([f'<div class="{sid}-item">{draw_image(test_data, x, class_id)}</div>'
-                                          for x in images_of_class_id])
+                thumbnail_images = [draw_image(img, labels) for img, labels in
```
Comment (noamzbr) [inline deepchecks/vision/checks/train_test_validation/new_labels.py:112]: What happens here if not .has_image?

File: deepchecks/vision/checks/train_test_validation/new_labels.py [hunk 3]
```diff
@@ -85,10 +54,32 @@ def __init__(
...
+            if data.task_type == TaskType.CLASSIFICATION:
+                self._update_images_dict(label, label.item(), image)
+            elif data.task_type == TaskType.OBJECT_DETECTION and len(label) > 0:
+                for class_id in np.unique(label[:, 0].numpy()):
+                    bbox_of_label = label[(label[:, 0] == class_id).nonzero().squeeze(1)]
+                    self._update_images_dict(bbox_of_label, class_id, image)
+            elif len(label) > 0:
+                raise DeepchecksValueError(f'Unsupported task type {data.task_type.value} for NewLabels check')
+
+    def _update_images_dict(self, label, class_id, image):
```
Comment (noamzbr) [inline deepchecks/vision/checks/train_test_validation/new_labels.py:77]: Should keep the indices for these images and return them as well, like in outliers.
Can be separate issue.

File: deepchecks/vision/checks/data_integrity/abstract_property_outliers.py [hunk 4]
```diff
@@ -57,36 +56,42 @@ class AbstractPropertyOutliers(SingleDatasetCheck):
...
-
+        self._draw_label_on_image = draw_label_on_image
         self._properties_results = None
 
     def initialize_run(self, context: Context, dataset_kind: DatasetKind):
         """Initialize the properties state."""
         data = context.get_data_by_kind(dataset_kind)
-
         self._properties_results = defaultdict(list)
-        # Take either alternative properties if defined or default properties defined by the child class
```
[the remark this answers is not in the document]
Comment (Nadav-Barak) [reply deepchecks/vision/checks/data_integrity/abstract_property_outliers.py:89]: I think the code is self explanatory

File: deepchecks/vision/checks/data_integrity/abstract_property_outliers.py [hunk 5]
```diff
@@ -95,79 +100,44 @@ def initialize_run(self, context: Context, dataset_kind: DatasetKind):
...
+        data = context.get_data_by_kind(dataset_kind)
         for prop_name, property_values in batch_properties.items():
             _ensure_property_shape(property_values, len(batch), prop_name)
+            images = data.batch_to_images(batch) if data.has_images else np.zeros(len(batch))
+            # TODO: fix when api refactor is complete
+            labels = data.batch_to_labels(batch) if data.has_labels else torch.zeros(len(batch))
+            # If the property or label is single value per image, wrap them in order to work on a fixed structure
+            if len(labels[0].shape) == 0:
+                labels = [l.resize(1, 1) for l in labels]
+            if not isinstance(property_values[0], t.Sequence):
```
[the remark this answers is not in the document]
Comment (Nadav-Barak) [reply deepchecks/vision/checks/data_integrity/abstract_property_outliers.py:112]: Changed if to use self.property_input_type

File: deepchecks/vision/checks/data_integrity/abstract_property_outliers.py [hunk 6]
```diff
@@ -216,31 +188,75 @@ def compute(self, context: Context, dataset_kind: DatasetKind) -> CheckResult:
...
-            The batch index of the sample to draw the image for.
-        index_of_value_in_sample : int
-            Each sample property is list, then this is the index of the outlier in the sample property list.
-        num_properties_in_sample
-            The number of values in the sample's property list.
-        """
-        pass
+        return CheckResult(check_result, display=display)
+
+    def _get_property_outlier_images(self, has_images: bool, lower_limit: float, property_name: str,
```
[the remark this answers is not in the document]
Comment (Nadav-Barak) [reply deepchecks/vision/checks/data_integrity/abstract_property_outliers.py:193]: OK

File: deepchecks/vision/checks/data_integrity/abstract_property_outliers.py [hunk 7]
```diff
@@ -216,31 +188,75 @@ def compute(self, context: Context, dataset_kind: DatasetKind) -> CheckResult:
...
+                    image_thumbnail = 'Image unavailable'
+                result.append((value, image_thumbnail))
+        return result
 
     @abstractmethod
     def get_default_properties(self, data: VisionData):
         """Return default properties to run in the check."""
         pass
 
+    def _update_lowest_highest_property_values(self, images: np.ndarray, labels: t.List, property_values: t.List,
```
[the remark this answers is not in the document]
Comment (Nadav-Barak) [reply deepchecks/vision/checks/data_integrity/abstract_property_outliers.py:225]: added

File: deepchecks/vision/checks/data_integrity/abstract_property_outliers.py [hunk 8]
```diff
@@ -57,36 +56,42 @@ class AbstractPropertyOutliers(SingleDatasetCheck):
...
     property_input_type: PropertiesInputType, default: PropertiesInputType.IMAGES
         The type of input to the properties, required for caching the results after first calculation.
-    n_show_top : int , default: 5
+    n_show_top : int , default: 3
         number of outliers to show from each direction (upper limit and bottom limit)
     iqr_percentiles: Tuple[int, int], default: (25, 75)
         Two percentiles which define the IQR range
     iqr_scale: float, default: 1.5
         The scale to multiply the IQR range for the outliers detection
+    draw_label_on_image: bool, default: True
```
[the remark this answers is not in the document]
Comment (Nadav-Barak) [reply deepchecks/vision/checks/data_integrity/abstract_property_outliers.py:65]: In case of image property outlier currently we do not draw bbox, and I think thats a valid choice

File: deepchecks/vision/checks/train_test_validation/new_labels.py (hunk 1 above)
Comment (Nadav-Barak) [reply deepchecks/vision/checks/train_test_validation/new_labels.py:74]: Currently we do not draw segmentation

Comment (Nadav-Barak) [reply deepchecks/vision/checks/train_test_validation/new_labels.py:74]: open an issue

File: deepchecks/vision/checks/train_test_validation/new_labels.py (hunk 2 above)
Comment (Nadav-Barak) [reply deepchecks/vision/checks/train_test_validation/new_labels.py:112]: same as now, we show an 'Image unavailble' where the image should be placed

**Review by noamzbr — APPROVED** [review]
